composition-js/src/merging/merge.ts

```diff
@@ -3606,4 +3689,344 @@ class Merger {
+        requirements.scopes = scopes;
+      }
+    }
+
+    if (this.policyDirective) {
+      const appliedDirective = element.appliedDirectivesOf(this.policyDirective)?.[0];
+      if (appliedDirective) {
+        const { policies } = appliedDirective.arguments();
+        requirements.policies = policies;
+      }
+    }
+
+    if (requirements.isAuthenticated || requirements.scopes || requirements.policies) {
+      return requirements;
+    } else {
+      return;
+    }
+  }
+
+  private verifyAuthRequirementsOnSelectionSet(authRequirements: AuthRequirements, selectionSet: SelectionSet) {
+    const parentType = this.schema.type(selectionSet.parentType.name);
+    for (const selection of selectionSet.selections()) {
+      if (selection instanceof FieldSelection) {
+        if (parentType instanceof InterfaceType) {
+          // if we are referencing an interface field we need to check against all implementations
+          for (const implType of parentType.possibleRuntimeTypes()) {
+            const requirementsOnImplType = this.authRequirementsOnElement(implType);
+            if (!authRequirements.satisfies(requirementsOnImplType)) {
+              const msg = `Field "${authRequirements.fieldCoordinate}" does not specify necessary @authenticated, @requiresScopes `
+                  + `and/or @policy auth requirements to access the transitive interface "${parentType}" data from ${authRequirements.directive} selection set.`;
+              throw ERRORS.MISSING_TRANSITIVE_AUTH_REQUIREMENTS.err(msg);
+            }
+            // re-check the implementation auth condition
+            this.verifyAuthOnFieldSelection(implType, selection, authRequirements);
+          }
+        }
+
+        if (parentType instanceof ObjectType) {
+          this.verifyAuthOnFieldSelection(parentType, selection, authRequirements);
+        }
+
+        if (selection.selectionSet) {
+          this.verifyAuthRequirementsOnSelectionSet(authRequirements, selection.selectionSet);
+        }
+      } else {
+        const condition = selection.element.typeCondition;
+        if (condition) {
+          const requirementsOnCondition = this.authRequirementsOnElement(condition);
+          if (!authRequirements.satisfies(requirementsOnCondition)) {
+            const msg = `Field "${authRequirements.fieldCoordinate}" does not specify necessary @authenticated, @requiresScopes `
+                + `and/or @policy auth requirements to access the transitive data in inline fragment type condition "${condition}" from ${authRequirements.directive} selection set.`;
+            throw ERRORS.MISSING_TRANSITIVE_AUTH_REQUIREMENTS.err(msg);
+          }
+        }
+        this.verifyAuthRequirementsOnSelectionSet(authRequirements, selection.selectionSet);
+      }
+    }
+  }
+
+  private verifyAuthOnFieldSelection(parentType: ObjectType, selection: FieldSelection, authRequirements: AuthRequirements) {
+    const field = parentType.field(selection.element.name);
```

Comment:
If we don't type-explode as mentioned above, this function shouldn't have to refetch its `field` from the merged schema (it also accordingly won't need to take in `parentType`).